### PR TITLE
Gtptunnel

### PR DESCRIPTION
**Pull request by prabinapattnaik**
## Summary

Added GTP Tunnel functionality in a pipelined.

## Test Plan

Default table 0 entry is added in a OVS:
cookie=0x0, duration=861.566s, table=0, n_packets=0, n_bytes=0, priority=0 actions=goto_table:1

Uplink and downlink entries are added in a OVS table 0:
cookie=0x0, duration=45.977s, table=0, n_packets=0, n_bytes=0, priority=10,tun_id=0x4,in_port=gtp0 actions=set_field:02:00:00:00:00:01->eth_src,set_field:ff:ff:ff:ff:ff:ff->eth_dst,set_field:0x7594587a00d->metadata,goto_table:1
cookie=0x0, duration=45.977s, table=0, n_packets=0, n_bytes=0, priority=10,ip,in_port=LOCAL,nw_dst=192.168.128.51 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x7594587a00d->metadata,goto_table:1
cookie=0x0, duration=45.977s, table=0, n_packets=0, n_bytes=0, priority=10,ip,in_port=mtr0,nw_dst=192.168.128.51 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x7594587a00d->metadata,goto_table:1

Uplink and downlink entries are deleted when tunnel was deleted

Written UT Suite for verify the UT:
"python/magma/pipelined/tests/test_gtp_flows.py"

## Additional Information

Forward data GTP tunnel and discard data GTP tunnel flow creation functionality also added.
[GTP_logs.zip](https://github.com/magma/magma/files/5041557/GTP_logs.zip)

### CODE REVIEW

File: fbinternal/cloud/go/go.sum [hunk 1]
```diff
@@ -492,6 +498,7 @@ golang.org/x/sys v0.0.0-20190403152447-81d4e9dc473e/go.mod h1:h1NjWce9XRLGQEsW7w
 golang.org/x/sys v0.0.0-20190422165155-953cdadca894/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20190502145724-3ef323f4f1fd h1:r7DufRZuZbWB7j439YfAzP8RPDa9unLkpwQKUYbIMPI=
 golang.org/x/sys v0.0.0-20190502145724-3ef323f4f1fd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/sys v0.0.0-20190507160741-ecd444e8653b/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20191010194322-b09406accb47 h1:/XfQ9z7ib8eEJX2hdgFTZJ/ntt0swNk5oYBziWeTCvY=
 golang.org/x/sys v0.0.0-20191010194322-b09406accb47/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd h1:xhmwyvizuTgC2qz7ZlMluP20uW+C3Rm0FD/WLDX8884=
```
Comment (pshelar) [inline fbinternal/cloud/go/go.sum:504]: this shld not be part of this patch.

File: fbinternal/cloud/go/services/fbinternal/servicers/exporter_servicer_test.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ func TestODSSubmit(t *testing.T) {
 		"",
 		"",
 		"magma",
+		"magma",
 		2,
```
Comment (pshelar) [inline fbinternal/cloud/go/services/fbinternal/servicers/exporter_servicer_test.go:45]: I am not sure why is this done.

File: lte/protos/smfupfif.proto [hunk 3]
```diff
@@ -0,0 +1,235 @@
+syntax = "proto3";
```
Comment (pshelar) [inline lte/protos/smfupfif.proto:1]: Please remove protobuf related changes from this patch. Lets just add GTP APP in this patch.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: fbinternal/cloud/go/go.sum (hunk 1 above)
Comment (prabinapattnaik) [reply fbinternal/cloud/go/go.sum:504]: I did not commit first few files (fbinternal/cloud/go/go.sum,  fbinternal/cloud/go/services/fbinternal/fbinternal/main.go,
fbinternal/cloud/go/services/fbinternal/servicers/exporter_servicer.go and fbinternal/cloud/go/services/fbinternal/servicers/exporter_servicer_test.go) and i think system was generated these files.
I will remove all these files

File: lte/protos/smfupfif.proto (hunk 3 above)
Comment (prabinapattnaik) [reply lte/protos/smfupfif.proto:1]: Yes. But in a UT test case, included the below message structures:

from lte.protos.smfupfif_pb2 import (
    SetGroupPDR,
    SetGroupFAR,
    PDI,
    LocalFTeid,
    ApplyAction,
    FwdParam,
    OuterHeaderCreation,
    SessionSet,
)
We are using these structures to passing parameters.

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+from magma.pipelined.openflow.magma_match import MagmaMatch
+from magma.pipelined.app.inout import INGRESS
+from ryu.lib.packet import ether_types
+from magma.pipelined.app.base import MagmaController, ControllerType
+#from lte.protos.smfupfif_pb2 import SetGroupPDR, SetGroupFAR
+
+GTP_PORT_MAC = "02:00:00:00:00:01"
+GTP = "gtp"
+
+class GtpFlows(MagmaController):
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/gtp_flows.py:24]: This is confusing name, it does process all ports in OVS bridge.
Lets call this classifier app,

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    def __init__(self, *args, **kwargs):
+        super(GtpFlows, self).__init__(*args, **kwargs)
+        self.config = self._get_config(kwargs['config'])
+        self.tbl_num = self._service_manager.get_table_num(self.APP_NAME)
+        self.next_table = self._service_manager.get_table_num(INGRESS)
+        self._uplink_port = OFPP_LOCAL
+        if self.config.mtr_ip:
+            self._mtr_service_enabled = True
+        else:
+            self._mtr_service_enabled = False
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/gtp_flows.py:47]: MTR is always on, so no need for config here.

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        self._install_default_tunnel_flows()
+
+    def _delete_all_flows(self):
+        flows.delete_all_flows_from_table(self._datapath, self.tbl_num)
+
+    def cleanup_on_disconnect(self, datapath):
+        self._delete_all_flows()
+
+    def convert_precedence_to_priority(self, precedence):
+        if precedence < flows.MAXIMUM_PRIORITY:
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/gtp_flows.py:84]: lets define type of each parameter.

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        
+        parser = self._datapath.ofproto_parser
+        priority = self.convert_precedence_to_priority(SetPDR.precedence)
+        # Add flow for gtp port
+        match = MagmaMatch(tunnel_id=SetPDR.pdi.local_f_teid.teid,
+                          in_port=self.config.gtp_port)
+
+        actions = [parser.OFPActionSetField(eth_src=GTP_PORT_MAC),
+                   parser.OFPActionSetField(eth_dst="ff:ff:ff:ff:ff:ff"),
+                   parser.OFPActionSetField(metadata=seid)]
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/gtp_flows.py:109]: AFAIK, seid is optional parameter, I think there shld be check for presence of this.

File: lte/gateway/python/magma/pipelined/openflow/flows.py [hunk 8]
```diff
@@ -103,7 +103,7 @@ def add_output_flow(datapath, table, match, actions=None, instructions=None,
 
 def add_flow(datapath, table, match, actions=None, instructions=None,
              priority=MINIMUM_PRIORITY, retries=3, cookie=0x0, idle_timeout=0,
-             hard_timeout=0):
+             hard_timeout=0, goto_table=0):
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/openflow/flows.py:106]: lets use None as value for skipping goto_table.

File: lte/protos/smfupfif.proto (hunk 3 above)
Comment (pshelar) [reply lte/protos/smfupfif.proto:1]: ok, let me think about this a bit

File: lte/gateway/python/magma/pipelined/openflow/flows.py [hunk 9]
```diff
@@ -135,16 +135,31 @@ def add_flow(datapath, table, match, actions=None, instructions=None,
...
+    if goto_table:
+        inst.append(parser.OFPInstructionGotoTable(goto_table))
+
     ryu_match = parser.OFPMatch(**match.ryu_match)
 
-    mod = parser.OFPFlowMod(datapath=datapath, priority=priority,
+    # cookie value is used for discard and forward data for GTP
+    if cookie > 0:
+        mod = parser.OFPFlowMod(datapath=datapath, priority=priority,
+                            match=ryu_match,table_id=table,
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/openflow/flows.py:158]: why do you need to check for cookie value.

File: lte/gateway/python/magma/pipelined/openflow/flows.py [hunk 10]
```diff
@@ -135,16 +135,31 @@ def add_flow(datapath, table, match, actions=None, instructions=None,
...
+                            out_group=ofproto.OFPG_ANY, out_port=ofproto.OFPP_ANY,
+                            cookie=cookie)
+    else:
+        mod = parser.OFPFlowMod(datapath=datapath, priority=priority,
                             match=ryu_match, instructions=inst,
                             table_id=table, cookie=cookie,
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/openflow/flows.py:164]: Can you fix your coding style config. I see some inconsistencies. for example the parameter list on second line should start from opening parentheses from line above.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py (hunk 5 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/app/gtp_flows.py:47]: OK

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py (hunk 4 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/app/gtp_flows.py:24]: I will do change the name.

File: lte/protos/smfupfif.proto (hunk 3 above)
Comment (prabinapattnaik) [reply lte/protos/smfupfif.proto:1]: Now I have modified the definition of install/delete GTP flows as a specific arguments like ue_ip addr, incoming_TEID, outgoing_TEID, EnodeB_ip addr, precedence, IMSI/SessionID for support 4G and 5G.
def _add_gtp_tunnel_flows(self, precedence:int, i_teid:int = None,
                              o_teid:int = None, ue_ip_adr:str = None,
                              enodeb_ip_addr:str = None, sid:int = None):

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py (hunk 6 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/app/gtp_flows.py:84]: I will do change.

File: lte/gateway/python/magma/pipelined/app/gtp_flows.py (hunk 7 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/app/gtp_flows.py:109]: Yes.Sure

File: lte/gateway/python/magma/pipelined/openflow/flows.py (hunk 8 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/openflow/flows.py:106]: OK

File: lte/gateway/python/magma/pipelined/openflow/flows.py (hunk 9 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/openflow/flows.py:158]: As OAI code, cookie value is used for discard data gtp tunnel and forward data gtp tunnel functionality.

File: lte/gateway/python/magma/pipelined/openflow/flows.py (hunk 10 above)
Comment (prabinapattnaik) [reply lte/gateway/python/magma/pipelined/openflow/flows.py:164]: Sure. I will fix it.
